once_test.py

```diff
@@ -167,25 +290,69 @@ def yielding_iterator():
         self.assertEqual(list(yielding_iterator()), [1, 2, 3])
         self.assertEqual(list(yielding_iterator()), [1, 2, 3])
 
+    def test_failing_generator(self):
+        counter = Counter()
+
+        @once.once
+        def sample_failing_fn():
+            yield counter.get_incremented()
+            raise ValueError("expected failure")
+
+        with self.assertRaises(ValueError):
+            list(sample_failing_fn())
+        with self.assertRaises(ValueError):
+            list(sample_failing_fn())
+        self.assertEqual(next(sample_failing_fn()), 1)
+        self.assertEqual(next(sample_failing_fn()), 1)
+
     def test_iterator_parallel_execution(self):
         counter = Counter()
+        counter.paused = True
 
         @once.once
         def yielding_iterator():
             nonlocal counter
             for _ in range(3):
                 yield counter.get_incremented()
 
-        with concurrent.futures.ThreadPoolExecutor(max_workers=16) as executor:
-            results = list(executor.map(lambda _: list(yielding_iterator()), range(32)))
+        with concurrent.futures.ThreadPoolExecutor(max_workers=_N_WORKERS) as executor:
+            results = executor.map(lambda _: list(yielding_iterator()), range(_N_WORKERS * 2))
+            counter.paused = False  # starter pistol, the race is off!
```

Comment:
I think better than this .paused would be a threading.Barrier, something like:

```python
barrier = threading.Barrier(_N_WORKERS * 2)

def execute(_):
    barrier.wait()
    return list(yielding_iterator())
results = executor.map(execute, range(_N_WORKERS * 2)) 
```
For a few reasons:
1. As is, paused is in the `main` thread, so there's no guarantee any of the other threads are ready, which is worse for synchronization (we still might get one thread that barrels through everything). The barrier means they're all ready.
2. We need to get out of the habit of implementing our own synchronization primitives (I'm guilty of this too with that asyncio.sleep(0)) :)

Reply:
done